batcheval: add MVCC-compliant `RevertRange` variant

erikgrinaker opened the pull request:
**roachpb: add `isAlone` for `RevertRangeRequest`**

Since `RevertRange` mutates MVCC history, we want them to be alone in a
batch. The DistSender will split any batches that have multiple such
requests.

Release note: None

**storage: add `NextKeyIgnoringTime()` for `MVCCIncrementalIterator`**

This patch adds a method `NextKeyIgnoringTime()` for
`MVCCIncrementalIterator`. This can be used to find the next key (as
opposed to version) of the iterator, ignoring the time bounds. It's
similar to `NextIgnoringTime()`, but calls `NextKey()` instead of
`Next()` on the underlying iterator.

Release note: None

**batcheval: add MVCC-compliant `RevertRange` variant**

This adds a new parameter `ExperimentalPreserveHistory` which, rather
than clearing keys above the target time, will write new values or
tombstones that reflect the state at the target time. For long runs of
new keys, this will instead drop an MVCC range tombstone. This makes the
command respect e.g. MVCC immutability, the closed timestamp, and
timestamp cache.

Note that MVCC range tombstones are currently experimental, and as such
this parameter is also experimental. Callers must call
`storage.CanUseExperimentalMVCCRangeTombstones()` before using it.

Resolves #70416.

Release note: None

erikgrinaker committed:
batcheval: add MVCC-compliant `RevertRange` variant

This adds a new parameter `ExperimentalPreserveHistory` to `RevertRange`
which, rather than clearing keys above the target time, will write new
values or tombstones that reflect the state at the target time. For long
runs of deleted keys, this will instead drop an MVCC range tombstone.
This makes the command respect e.g. MVCC immutability, the closed
timestamp, and timestamp cache.

Note that MVCC range tombstones are currently experimental, and as such
this parameter is also experimental. Callers must call
`storage.CanUseExperimentalMVCCRangeTombstones()` before using it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76478)
<!-- Reviewable:end -->

dt commented on pkg/roachpb/api.go:
```diff
@@ -1263,7 +1263,13 @@ func (*ClearRangeRequest) flags() flag { return isWrite | isRange | isAlone }
+	if rrr.ExperimentalPreserveHistory {
+		return isRead | isWrite | isRange | isAlone | updatesTSCache | appliesTSCache
+	}
+	// TODO(erikgrinaker): add isAlone?
```
I think so, yes

dt commented on pkg/roachpb/api.proto:
```diff
@@ -402,18 +402,34 @@ message ClearRangeResponse {
   // for that time are still there — or the request will fail.
   util.hlc.Timestamp target_time = 2 [(gogoproto.nullable) = false];
 
+  // ExperimentalPreserveHistory will preserve MVCC history by, rather than
```
We _could_ decouple these, and say 1) PreserveHistory will do the tombstone/undo, and is thus mvcc and 2) experimental_mvcc_delrange allows preserve history, if in use, to use cheap range deletion tombstones rather than slow/expensive individual tombstones. Then we can do mvcc but just slow without being in experimental territory?

erikgrinaker replied:
We could, but that would require more testing and such, and I can't afford that right now. Any reason to believe we would use this in 22.1?

dt replied:
nope, works for me.

dt commented on pkg/storage/mvcc_incremental_iterator.go:
```diff
@@ -533,6 +533,33 @@ func (i *MVCCIncrementalIterator) NextIgnoringTime() {
 	}
 }
 
+// NextKeyIgnoringTime returns the next distinct key that would be encountered
```
should this be in one of the prior commits? I guess waiting to add it here gets it test coverage.

erikgrinaker replied:
I could split it out and add some unit tests. That's probably a good idea anyway, since this could get picked up for production usage somewhere.

dt reviewed (commented):
Just one API question from me

erikgrinaker committed:
roachpb: add `isAlone` for `RevertRangeRequest`

Since `RevertRange` mutates MVCC history, we want them to be alone in a
batch. The DistSender will split any batches that have multiple such
requests.

Release note: None

erikgrinaker committed:
storage: add `NextKeyIgnoringTime()` for `MVCCIncrementalIterator`

This patch adds a method `NextKeyIgnoringTime()` for
`MVCCIncrementalIterator`. This can be used to find the next key (as
opposed to version) of the iterator, ignoring the time bounds. It's
similar to `NextIgnoringTime()`, but calls `NextKey()` instead of
`Next()` on the underlying iterator.

Release note: None

aliher1911 commented on pkg/storage/mvcc.go:
```diff
@@ -2218,6 +2274,211 @@ func recordIteratorStats(traceSpan *tracing.Span, iteratorStats IteratorStats) {
+		return nil, nil
+	}
+
+	// We set up an ingremental iterator from the revert time to look for any
```
incremental

aliher1911 commented on pkg/storage/mvcc.go:
```diff
@@ -2218,6 +2274,211 @@ func recordIteratorStats(traceSpan *tracing.Span, iteratorStats IteratorStats) {
+	var revertKey roachpb.Key
+	var revertValue, revertValueFrom []byte
+	iter.SeekGE(MVCCKey{Key: startKey})
+	for {
```
I found following mvcc iterator loops where we iterate versions and keys and feed it to "state machine" pretty hard to follow. Maybe the better idea would be to nest iteration of versions into iteration of keys and only track deletions inbetween?
We may leave it as a todo since we have some deadlines.

aliher1911 reviewed (commented):
Mostly nits about comments, but I think there's some room for improvement in ExperimentalMVCCRevertRange which we should address before in becomes non-experimental.

erikgrinaker commented on pkg/storage/mvcc_incremental_iterator.go:
```diff
@@ -518,6 +533,34 @@ func (i *MVCCIncrementalIterator) NextIgnoringTime() {
 }
 
+// NextKeyIgnoringTime returns the next distinct key that would be encountered
+// in a non-incremental iteration by moving the underlying non-TBI iterator
```
Updated.

dt reviewed (approved):
API as it'll be available to bulk/sql clients LGTM!

erikgrinaker commented:
TFTRs!

bors r=aliher1911,dt

erikgrinaker commented:
Some weird agent failure for `git checkout`.

bors retry

erikgrinaker commented:
Unrelated flake. Third time's the charm!

bors retry
